edb/server/protocol/ai_ext.py

```diff
@@ -0,0 +1,1305 @@
+)
+
+import asyncio
+import contextvars
+import http
+import itertools
+import json
+import logging
+
+import httpx
+import httpx_sse
+
+from edb import errors
+from edb.common import debug
+from edb.common import markup
+from edb.common import uuidgen
+
+from edb.server import compiler
+from edb.server.compiler import sertypes
+from edb.server.protocol import execute
+
+if TYPE_CHECKING:
+    from edb.server import dbview
+    from edb.server import tenant as srv_tenant
+    from edb.server import pgcon
+    from edb.server.protocol import protocol
+
+
+logger = logging.getLogger("edb.server.ai_ext")
+
+
+def start_extension(
+    tenant: srv_tenant.Tenant,
+    dbname: str,
+) -> None:
+    task_name = _get_builder_task_name(dbname)
+    task = tenant.get_task(task_name)
+    if task is None:
+        logger.info(f"starting AI extension tasks on database {dbname!r}")
+        tenant.create_task(
+            _ext_ai_index_builder_controller_loop(tenant, dbname),
+            interruptable=True,
+            name=task_name,
+        )
+
+
+def stop_extension(
+    tenant: srv_tenant.Tenant,
+    dbname: str,
+) -> None:
+    task_name = _get_builder_task_name(dbname)
+    task = tenant.get_task(task_name)
+    if task is not None:
+        logger.info(f"stopping AI extension tasks on database {dbname!r}")
+        task.cancel()
+
+
+def _get_builder_task_name(dbname: str) -> str:
+    return f"ext::ai::index builder on database {dbname!r}"
+
+
+_task_name = contextvars.ContextVar(
+    "ext_ai_index_builder_task_name", default="-")
+
+
+async def _ext_ai_index_builder_controller_loop(
+    tenant: srv_tenant.Tenant,
+    dbname: str,
+) -> None:
+    task_name = _get_builder_task_name(dbname)
+    _task_name.set(task_name)
+    logger.info(f"started {task_name}")
+    db = tenant.get_db(dbname=dbname)
+    holding_lock = False
+    try:
+        while True:
+            try:
+                pgconn = await tenant.acquire_pgcon(dbname)
+                models = []
+                processed = 0
+                try:
+                    models = await _ext_ai_fetch_active_models(pgconn)
+                    if models:
+                        if not holding_lock:
+                            holding_lock = await _ext_ai_lock(pgconn)
+                        if holding_lock:
+                            try:
+                                processed = await _ext_ai_index_builder_work(
+                                    db, pgconn, models)
+                            finally:
+                                if processed == 0:
+                                    await _ext_ai_unlock(pgconn)
+                                    holding_lock = False
+                finally:
+                    tenant.release_pgcon(dbname, pgconn)
+            except Exception:
+                logger.exception(f"caught error in {task_name}")
+
+            if processed == 0:
+                # No work, sleep for a bit.
+                await asyncio.sleep(10)
```

Comment:
Move this to server.defines?

Reply:
This got replaced with a config var instead.